### PR TITLE
added heffte-magma example

### PR DESCRIPTION
**Pull request by mkstoyanov**
* added heffte-magma example

TODO:
- [x] Implement (in CMake) testing for the new example
- [x] Document additions in the existing or new `README.md` files including the main `README.md`

### CODE REVIEW

**Comment by mkstoyanov** [comment]
I guess someone else should take a look at this and approve if OK.

**Review by stomov — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in branch [testing-v0.4.0](https://github.com/xsdk-project/xsdk-examples/tree/testing-v0.4.0) for testing ...

File: heffte/CMakeLists.txt
Comment (v-dobrev) [inline heffte/CMakeLists.txt:1]: To be consistent with other packages, can you please move the `find_package` command in the top-level CMake file?

Comment (luszczek) [reply heffte/CMakeLists.txt:1]: What I see that `heFFTe` is enabled by default and `MAGMA` is disabled by default. Maybe we need logic that enables examples based on a mix of options that were enabled in top `CMakeLists.txt`.

Comment (mkstoyanov) [reply heffte/CMakeLists.txt]: The if-statement guarding this is in the main CMakeLists.txt and by default heFFTe is not enabled and the entire subdirectory will not be included.

But @luszczek has a good point, packages are set to on/off seemingly at random. We should have some better criteria.

Comment (mkstoyanov) [reply heffte/CMakeLists.txt]: Moved the `find_package()`

Comment (balay) [reply heffte/CMakeLists.txt]: merged to `testing-v0.4.0` branch

**Comment by mkstoyanov** [comment]
I will have to investigate here, there may be a bug in @develop for heFFTe. Not sure about the release.

File: heffte/CMakeLists.txt
Comment (v-dobrev) [reply heffte/CMakeLists.txt]: > The if-statement guarding this is in the main CMakeLists.txt and by default heFFTe is not enabled and the entire subdirectory will not be included.
> 
> But @luszczek has a good point, packages are set to on/off seemingly at random. We should have some better criteria.

I thought the logic here is that this example can work both with and without MAGMA and that dependency is handled through heFFTe, so there's no need for any explicit logic regarding MAGMA in this example. Is that not correct?

Comment (mkstoyanov) [reply heffte/CMakeLists.txt]: I don't think this was in regard the heFFTe-MAGAM integration but a statement regarding all packages. Right now, some packages are enabled by default and others not, seemingly a bit at random.

Comment (v-dobrev) [reply heffte/CMakeLists.txt]: Okay, I see. I guess the decision about the default state of a package, enabled or disabled, was made by the person that added the specific package. So, yes, it is a bit random, I guess.

**Comment by mkstoyanov** [comment]
Should be fixed in `@develop` after this is merged:

https://github.com/icl-utk-edu/heffte/pull/23

**Comment by mkstoyanov** [comment]
@balay can you try testing `@develop` again?

**Comment by balay** [comment]
heffte@develop  works (with cuda).  With just HEFFTE, PETSC enabled - I get:
```
balay@compute-386-08:/scratch/balay/spack$ cat /scratch/balay/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.3.0/xsdk-examples-0.4.0-qc4huhjzaqohlbbiw4tqdmaboeqmt45m/.spack/install-time-test-log.txt
==> Testing package xsdk-examples-0.4.0-qc4huhj
==> [2023-05-10-15:47:52.227555] Running build-time tests
==> [2023-05-10-15:47:52.227670] RUN-TESTS: build-time tests [check]
==> [2023-05-10-15:47:52.228280] '/scratch/balay/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.3.0/cmake-3.26.3-64hpouabequlexvshbbhjgeqbdkoftou/bin/ctest' '--output-on-failure'
Test project /scratch/balay/spack/spack-stage/spack-stage-xsdk-examples-0.4.0-qc4huhjzaqohlbbiw4tqdmaboeqmt45m/spack-build-qc4huhj
    Start 1: heFFTe-MAGMA
1/3 Test #1: heFFTe-MAGMA .....................   Passed    5.27 sec
    Start 2: PETSc-ex19_1
2/3 Test #2: PETSc-ex19_1 .....................   Passed    0.74 sec
    Start 3: PETSc-ex19_cuda
3/3 Test #3: PETSc-ex19_cuda ..................   Passed    4.80 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =  10.81 sec
```

However we can't use @develop for spack install of xsdk-examples@0.4.0  - so needs to be disabled here? cc: @v-dobrev 

But then - the non-cuda build works fine with heffte@2.3.0
```
balay@compute-386-08:/scratch/balay/spack.x$ cat /scratch/balay/spack.x/opt/spack/linux-ubuntu22.04-icelake/gcc-11.3.0/xsdk-examples-0.4.0-r55k4rm2t23gjypeftq5rmhl4vnhtec3/.spack/install-time-test-log.txt
==> Testing package xsdk-examples-0.4.0-r55k4rm
==> [2023-05-10-15:52:51.481390] Running build-time tests
==> [2023-05-10-15:52:51.481586] RUN-TESTS: build-time tests [check]
==> [2023-05-10-15:52:51.482544] '/scratch/balay/spack.x/opt/spack/linux-ubuntu22.04-icelake/gcc-11.3.0/cmake-3.26.3-64hpouabequlexvshbbhjgeqbdkoftou/bin/ctest' '--output-on-failure'
Test project /scratch/balay/spack.x/spack-stage/spack-stage-xsdk-examples-0.4.0-r55k4rm2t23gjypeftq5rmhl4vnhtec3/spack-build-r55k4rm
    Start 1: heFFTe-MAGMA
1/2 Test #1: heFFTe-MAGMA .....................   Passed    0.74 sec
    Start 2: PETSc-ex19_1
2/2 Test #2: PETSc-ex19_1 .....................   Passed    1.42 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =   2.17 sec
```

**Comment by mkstoyanov** [comment]
> can you paste the diff here?

```
diff --git a/cmake/HeffteConfig.cmake b/cmake/HeffteConfig.cmake
index 8bbe907..df1cf86 100644
--- a/cmake/HeffteConfig.cmake
+++ b/cmake/HeffteConfig.cmake
@@ -51,7 +51,7 @@ if (@Heffte_ENABLE_MAGMA@ AND NOT TARGET Heffte::MAGMA)
     set_target_properties(Heffte::MAGMA PROPERTIES INTERFACE_INCLUDE_DIRECTORIES @HeffteMAGMA_INCLUDES@)
     if (@Heffte_ENABLE_CUDA@)
         list(FILTER CUDA_CUBLAS_LIBRARIES EXCLUDE REGEX "-NOTFOUND$") # work-around CMake 3.10 + CUDA 10
-        target_link_libraries(Heffte Heffte::MAGMA INTERFACE ${CUDA_CUBLAS_LIBRARIES})
+        target_link_libraries(Heffte::MAGMA INTERFACE ${CUDA_CUBLAS_LIBRARIES})
     endif()
     if (@Heffte_ENABLE_ROCM@)
         find_package(rocblas REQUIRED)
```

**Comment by balay** [comment]
Thanks! pushed to `packages/xsdk-examples-v0.4.0` branch at https://github.com/spack/spack/commit/e4fc78d85ab30f4baa97b5a579b1102c9c436562

**Comment by mkstoyanov** [comment]
Could you please rename the patch file to `cmake-magma-v230.patch` Having the version in the patch name helps keep track of things.

Thanks!

**Comment by balay** [comment]
ok renamed.

https://github.com/spack/spack/commit/7bbe4832aabce422eec0cf3f033ade4113b4da44

And to update now the xsdk-examples+cuda test works [with petsc, heffte only] enabled.

File: heffte/CMakeLists.txt
Comment (v-dobrev) [inline heffte/CMakeLists.txt:4]: Use full executable path -- needed when `.` is not in the `PATH` and `mpiexec` does not find it.
```suggestion
xsdk_add_test(NAME heFFTe-MAGMA MPI_NPROCS 2 COMMAND $<TARGET_FILE:heffte_example>)
```

**Pull request merged**
